File: leetcode/2. Add Two Numbers/Concise/concise.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	l3 := &ListNode{}
	head := l3
	for sum := 0; l1 != nil || l2 != nil || sum > 0; {
		if l1 != nil {
			sum += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			sum += l2.Val
			l2 = l2.Next
		}
		l3.Next = &ListNode{Val: sum % 10}
		l3 = l3.Next
		sum /= 10
	}
	return head.Next
}

File: leetcode/2. Add Two Numbers/Concise/concise.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* l3 = new ListNode();
        ListNode* head = l3;
        for(int carry=0; l1 != NULL || l2 != NULL || carry > 0; carry /= 10){
            if (l1 != NULL){
                carry += l1->val;
                l1 = l1->next;
            }
            if (l2 != NULL){
                carry += l2->val;
                l2 = l2->next;
            }
            l3->next = new ListNode(carry%10);
            l3 = l3->next;
        }
        return head->next;
    }
};